### DISCUSSION TITLE
stm32 subghz DT

### DISCUSSION THREAD

**Discussion by nguilbaudesoft** [zephyrproject-rtos/zephyr#36512] [Q&A]
Hello,

We are trying to use the STM32WL which includes the LoRA radio inside.

In the DT of the STM32WL there is a mention of the subghz SPI, but this SPI doesn't use the specific SUBGHZ HAL in the STM32 lib. Could you confirm that the STM32 SPI driver will work with for discussing with the internal LoRA tranceiver inside the STM32WL ?

in https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/wl/stm32wl.dtsi :
		subghzspi: spi@58010000 {
			compatible = **"st,stm32-spi-fifo", "st,stm32-spi"**;
			#address-cells = <1>;
			#size-cells = <0>;
			reg = <0x58010000 0x400>;
			interrupts = <44 5>;
			clocks = <&rcc STM32_CLOCK_BUS_APB3 0x00000001>;
			status = "disabled";
			label = "SUBGHZSPI";
		};

What about the control signal: clock, gpios between the STM32 and LoRA tranceiver ?

Thanks,
Nicolas.

**Comment by erwango** [comment]
Thanks for pointing at this.

There are 2 possible implementations to address LORA support for STM32WL:
- Using SubGHz HAL: This will require the implementation of a specific SubGHz HAL wrapper, HAL functions will take care of SPI 
  communication, C2 configuration, .. What will be left to Zephyr configuration (MSP init) is to configure IRQ, clocks with usual API.
  In the end, this should be somewhat close to Zephyr STM32 Ethernet driver...
- W/o use of SubGHz HAL: This requires more investigation. But from what I can see SUBGHz SPI is anyway different from usual SPI, so 
  I don't think Zephyr STM32 SPI Driver will work as is.

First option will likely have some small conflicts with usual Zephyr APIs to deal with (exti for instance), but that shouldn't be too much trouble in the end.
The second option should allow to avoid this conflicts, but would require more development, w/o the benefit of re-using SPI driver.

Is that helping ?

**Comment by erwango** [comment]
Adding @martinjaeger, @Mani-Sadhasivam, @tagunil, @str4t0m, @fabiobaltieri who also seem interested in WL, and maybe already had a look as well

**Reply by fabiobaltieri** [reply]
Hey thanks, I have a working LoRa patch for the WL using the unmodified Zephyr SPI driver and some HAL mod for the sx1261 driver to handle the control signals (which are NOT through GPIOs). Seems to work fine, will upload it over the weekend for discussion.

**Reply by erwango** [reply]
@fabiobaltieri That's an interesting alternative, txs!

**Reply by fabiobaltieri** [reply]
Here it is: https://github.com/zephyrproject-rtos/zephyr/pull/36573.

**Comment by erwango** [comment]
@nguilbaudesoft, #36573 just got merged.

**Comment by plmwd** [comment]
Are there any other options needed to get the lora e5 dev board to receive correctly? The timing seems to be off, or something else is happening - going off the class_a sample. I can send, but I mostly get errors receiving. I imagine it's the receive window timings, correct? Any help here would be greatly appreciated! I'm really loving zephyr, but I'm still trying to get comfortable with the code base. 😃
